src/types/relocatable.rs

```diff
@@ -88,43 +88,82 @@ impl Display for Relocatable {
 }
 
 impl Add<usize> for Relocatable {
-    type Output = Relocatable;
-    fn add(self, other: usize) -> Self {
-        relocatable!(self.segment_index, self.offset + other)
+    type Output = Result<Relocatable, MathError>;
+    fn add(self, other: usize) -> Result<Self, MathError> {
+        self.offset
+            .checked_add(other)
+            .map(|x| Relocatable::from((self.segment_index, x)))
+            .ok_or(MathError::RelocatableAddUsizeOffsetExceeded(self, other))
     }
 }
 
+/// Warning: may panic if self.offset + rhs exceeds usize::MAX
 impl AddAssign<usize> for Relocatable {
     fn add_assign(&mut self, rhs: usize) {
         self.offset += rhs
     }
 }
 
 impl Add<i32> for Relocatable {
-    type Output = Relocatable;
-    fn add(self, other: i32) -> Self {
+    type Output = Result<Relocatable, MathError>;
+    fn add(self, other: i32) -> Result<Self, MathError> {
         if other >= 0 {
-            relocatable!(self.segment_index, self.offset + other as usize)
+            self + other as usize
         } else {
-            relocatable!(
-                self.segment_index,
-                self.offset - other.unsigned_abs() as usize
-            )
+            self - other.unsigned_abs() as usize
         }
     }
 }
+impl Add<&Felt> for Relocatable {
+    type Output = Result<Relocatable, MathError>;
+    fn add(self, other: &Felt) -> Result<Relocatable, MathError> {
+        let big_offset = other + self.offset;
+        let new_offset = big_offset
+            .to_usize()
+            .ok_or_else(|| MathError::RelocatableAddFeltOffsetExceeded(self, other.clone()))?;
+        Ok(Relocatable {
+            segment_index: self.segment_index,
+            offset: new_offset,
+        })
+    }
+}
 
-impl Add<i32> for &Relocatable {
-    type Output = Relocatable;
-    fn add(self, other: i32) -> Relocatable {
-        if other >= 0 {
-            relocatable!(self.segment_index, self.offset + other as usize)
-        } else {
-            relocatable!(
-                self.segment_index,
-                self.offset - other.unsigned_abs() as usize
-            )
+/// Adds a MaybeRelocatable to self
+/// Cant add two relocatable values
+impl Add<&MaybeRelocatable> for Relocatable {
+    type Output = Result<Relocatable, MathError>;
+    fn add(self, other: &MaybeRelocatable) -> Result<Relocatable, MathError> {
```

Comment:
Typical (equivalent) signature:
```suggestion
    type Output = Result<Relocatable, MathError>;
    fn add(self, other: &MaybeRelocatable) -> Self::Output {
```

I'm not 100% sure returning a `Result<_>` from an operator is what anybody would expect, as typically they'd use `checked_*` to handle errors and let it `panic!` otherwise. Still, this is infinitely better than what we had, so I'm OK with it.